integration-tests/src/test/scala/za/co/absa/spline/BasicIntegrationTests.scala

```diff
@@ -117,15 +122,41 @@ class BasicIntegrationTests extends AsyncFlatSpec
               .write.mode(Append).saveAsTable(tableName)
           }
 
-          (plan2, _) <- captor.lineageOf {
+          // Spark 3.4+ is creating 2 commands for both writes here so we need to ignore one
+          // We only want the one that is from CreateDataSourceTableAsSelectCommand
+          // The one we ignore here is an extra InsertIntoHadoopFsRelationCommand
+          // They can come out of order so we need to filter out which one is which.
+          (plan2, _) <- if (ver"$SPARK_VERSION" >= ver"3.4.0") {
```

Comment:
I am not a fan of taking this approach but this does allow the tests to still pass - if we are comfortable with knowing that Spark is firing additional events here we get the same behavior.

Reply:
So this is what's confusing me a little bit - we get both a CreateDataSourceTableAsSelectCommand & InsertIntoHadoopFsRelationCommand (printed from LineageHarvester) in an a single Spark Action (and they aren't guaranteed to appear in the same order as well, hence why I needed to do that weird filter to find the right one).

I think I'm going to run a custom OutputQueryListener and see if Spark itself happens to print out two actions as well. If not I'm going to be even more confused 😅

Reply:
But yes - this didn't happen in < 3.4

Reply:
Ok I think I've just confirmed that this is indeed Spark and not Spline doing this. I made a listener specifically for this test like so:

```scala
class TestListener extends QueryExecutionListener {

  override def onSuccess(funcName:  String, qe:  QueryExecution, durationNs:  Long): Unit = {
    println("A COMMAND JUST RAN")
    println(qe.commandExecuted.getClass.getCanonicalName)
  }
  override def onFailure(funcName:  _root_.scala.Predef.String, qe:  _root_.org.apache.spark.sql.execution.QueryExecution, exception:  scala.Exception): Unit = {

  }
}
```

3.3 tests give just per CTAS action.

```
A COMMAND JUST RAN
org.apache.spark.sql.execution.command.CreateDataSourceTableAsSelectCommand
```

But 3.4 tests give this per CTAS action:

```
A COMMAND JUST RAN
org.apache.spark.sql.execution.datasources.InsertIntoHadoopFsRelationCommand
A COMMAND JUST RAN
org.apache.spark.sql.execution.command.CreateDataSourceTableAsSelectCommand
```

This seems to support what I saw from printing out in LineageHarvester.

I unfortunately don't have the context to say whether this is ok or not :sweat_smile: . I would imagine Spline UI users would see this additional event, no?

The correct lineage events are getting generated but there's extra noise. I know InsertIntoHadoopFsRelation is associated with other writes occurring so your theory of Spark changing something under the hood is probably correct.

Reply:
I think I found the [Spark PR](https://github.com/apache/spark/commit/4b40920e33176fc8b18380703e4dcf4d16824094) that introduced this behavior change in 3.4 - TLDR v1 data writes originally in CTAS turned into the two relation's we're seeing here under the two types of CTAS supported. They call out command nesting in various places.

Reply:
So you are sure that both of those commands are triggered by `.write.mode(Append).saveAsTable(tableName)`?

This is not a question for testing only, but generally how to handle this in the application. Usually there is only one lineage for one write. The simplest solution would be to ignore one of them, but how to check if what we are ignoring is actually duplication?

Reply:
I read the Spark ticket, they actually separate the action of table creation and data insert, that is not a problem. We could generate events for both, but it should be clear that no data are inserted in the table creation lineage.

Reply:
Would there be another suite/case that we'd want for something like this? I suppose that might depend on the actual events that Spline would be outputting here now.